src/algorithms/telemetry_decoder/gnuradio_blocks/beidou_b2a_telemetry_decoder_cc.cc

```diff
@@ -0,0 +1,448 @@
+    bool dump) : gr::block("beidou_b2a_telemetry_decoder_cc", gr::io_signature::make(1, 1, sizeof(Gnss_Synchro)),
+                     gr::io_signature::make(1, 1, sizeof(Gnss_Synchro)))
+{
+    // Ephemeris data port out
+    this->message_port_register_out(pmt::mp("telemetry"));
+    // initialize internal vars
+    d_dump = dump;
+    d_satellite = Gnss_Satellite(satellite.get_system(), satellite.get_PRN());
+    LOG(INFO) << "Initializing BEIDOU B2a TELEMETRY DECODING";
+    // Define the number of sampes per symbol. Notice that BEIDOU has 2 rates, !!!Change
+    //one for the navigation data and the other for the preamble information
+    d_samples_per_symbol = (BEIDOU_B2a_CODE_RATE_HZ / BEIDOU_B2a_CODE_LENGTH_CHIPS) / BEIDOU_B2a_SYMBOL_RATE_BPS;
+
+    // Set the preamble information
+    unsigned short int preambles_bits[BEIDOU_CNAV2_PREAMBLE_LENGTH_BITS] = BEIDOU_CNAV2_PREAMBLE;
+    // Since preamble rate is different than navigation data rate we use a constant
+    d_symbols_per_preamble = BEIDOU_CNAV2_PREAMBLE_LENGTH_SYMBOLS;
+
+    memcpy(static_cast<unsigned short int *>(this->d_preambles_bits), static_cast<unsigned short int *>(preambles_bits), BEIDOU_CNAV2_PREAMBLE_LENGTH_BITS * sizeof(unsigned short int));
+
+    // preamble bits to sampled symbols
+    d_preambles_symbols = static_cast<signed int *>(malloc(sizeof(signed int) * d_symbols_per_preamble));
+    int n = 0;
+    for (int i = 0; i < BEIDOU_CNAV2_PREAMBLE_LENGTH_BITS; i++)
+        {
+            for (unsigned int j = 0; j < BEIDOU_CNAV2_TELEMETRY_SYMBOLS_PER_PREAMBLE_BIT; j++)
+                {
+                    if (d_preambles_bits[i] == 1)
+                        {
+                            d_preambles_symbols[n] = 1;
+                        }
+                    else
+                        {
+                            d_preambles_symbols[n] = -1;
+                        }
+                    n++;
+                }
+        }
+    d_sample_counter = 0;
+    d_stat = 0;
+    d_preamble_index = 0;
+
+    d_flag_frame_sync = false;
+
+    d_flag_parity = false;
+    d_TOW_at_current_symbol = 0;
+    Flag_valid_word = false;
+    delta_t = 0;
+    d_CRC_error_counter = 0;
+    d_flag_preamble = false;
+    d_channel = 0;
+    flag_TOW_set = false;
+    d_preamble_time_samples = 0;
+}
+
+
+beidou_b2a_telemetry_decoder_cc::~beidou_b2a_telemetry_decoder_cc()
+{
+    delete d_preambles_symbols;
+    if (d_dump_file.is_open() == true)
+        {
+            try
+                {
+                    d_dump_file.close();
+                }
+            catch (const std::exception &ex)
+                {
+                    LOG(WARNING) << "Exception in destructor closing the dump file " << ex.what();
+                }
+        }
+}
+
+
+void beidou_b2a_telemetry_decoder_cc::decode_string(double *frame_symbols, int frame_length)
+{
+    double chip_acc = 0.0;
+    int chip_acc_counter = 0;
+
+    // 1. Transform from symbols to bits
```

Comment:
All lines below this seems to be for GLONASS Telemetry Decoding and not for Beidou. This seems wrong